scrapy_autoextract/providers.py

```diff
@@ -1,114 +1,226 @@
+class AutoExtractProvider(PageObjectInputProvider):
+    """Provider for AutoExtract data"""
+    page_type_class: ClassVar[Type]
 
-    @classmethod
-    def register(cls):
-        """Register this provider for its provided class on scrapy-poet
-        registry. This will make it possible to declare provided class as
-        a callback dependency when writing Scrapy spiders.
-        """
-        register(cls, cls.provided_class)
+    # pageType requested when only html is required
+    page_type_class_for_html: ClassVar[AutoExtractData] = AutoExtractProductData
+    html_query_attribute: ClassVar[str] = "fullHtml"
 
     @classmethod
-    def get_page_type(cls) -> str:
-        """Page type is defined by the class attribute `item_key` available on
-        `autoextract_poet.page_inputs` classes.
-        """
-        return cls.provided_class.item_key
+    def provided_classes(cls, type_: Callable) -> bool:
+        return (inspect.isclass(type_)
+                and issubclass(type_, (AutoExtractData, AutoExtractHtml)))
 
+    def __init__(self, crawler: Crawler):
+        """Initialize provider storing its dependencies as attributes."""
+        self.crawler = crawler
+        self.settings = crawler.spider.settings
+        self.logger = self.crawler.spider.logger
+        self.task_manager = get_autoextract_task_manager(crawler)
+        self.aiohttp_session = None
+        self.crawler.signals.connect(self.close_aiohttp_session,
+                                     signal=signals.spider_closed)
+        retries_count = self.settings.getint(
+                "AUTOEXTRACT_MAX_QUERY_ERROR_RETRIES", 3)
+        self.common_request_kwargs = dict(
+            api_key=self.settings.get("AUTOEXTRACT_USER"),
+            endpoint=self.settings.get("AUTOEXTRACT_URL"),
+            max_query_error_retries=retries_count,
+            session=self.aiohttp_session
+        )
+        per_domain_concurrency = get_concurrent_requests_per_domain(self.settings)
+        self.per_domain_semaphore = SlotsSemaphore(per_domain_concurrency)
+        self.logger.info(
+            f"AutoExtractProvider started. Retries: {retries_count}, "
+            f"per domain concurrency: {per_domain_concurrency}"
+        )
 
-class ArticleDataProvider(_Provider):
+    async def create_aiohttp_session(self) -> aiohttp.ClientSession:
+        concurrent_connections = self.settings.getint("CONCURRENT_REQUESTS", 16)
+        self.logger.info(
+            f"AutoExtractProvider concurrent requests: {concurrent_connections}"
+        )
+        return create_session(connection_pool_size=concurrent_connections)
 
-    provided_class = AutoExtractArticleData
+    async def close_aiohttp_session(self):
+        if self.aiohttp_session:
+            await self.aiohttp_session.close()
 
+    def create_retry_wrapper(self):
+        return RetryFactory().build()
 
-class ProductDataProvider(_Provider):
+    async def do_request(self, *args, **kwargs):
```

Comment:
Not sure getting it. `async` and `await` are required as `request_raw` is a coroutine. `args` and `kwargs` are also required as they are carrying arguments (they are not empty). But maybe you are meaning a different thing?

I think that maybe you are referring that given that `request_raw` returns an `awaitable` when invoked and so must do `do_request`, that we can just omit the `await` and the `async`. Interesting, good point. In any case, I think is more clear with the `async` and `await` so that it is explicit that everything is async.

Reply:
Yeah, that was my idea here.
Since it already returns the coroutine, we don't need to make it async and await.
Though, thinking now, it makes sense to make it explicit and easier to understand the API..